gax/src/main/java/com/google/api/gax/tracing/OpencensusTracer.java

```diff
@@ -0,0 +1,386 @@
+    Map<String, AttributeValue> attributes = baseOperationAttributes();
+
+    span.putAttributes(attributes);
+    span.end();
+  }
+
+  @Override
+  public void operationCancelled() {
+    Map<String, AttributeValue> attributes = baseOperationAttributes();
+    span.putAttributes(attributes);
+    span.end(
+        EndSpanOptions.builder()
+            .setStatus(Status.CANCELLED.withDescription("Cancelled by caller"))
+            .build());
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void operationFailed(Throwable error) {
+    Map<String, AttributeValue> attributes = baseOperationAttributes();
+
+    span.putAttributes(attributes);
+    span.end(EndSpanOptions.builder().setStatus(convertErrorToStatus(error)).build());
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void connectionSelected(int id) {
+    span.addAnnotation(
+        "Connection selected", ImmutableMap.of("id", AttributeValue.longAttributeValue(id)));
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void attemptStarted(int attemptNumber) {
+    currentAttemptId = attemptNumber;
+    attemptRequests = 0;
+    attemptResponses = 0;
+
+    HashMap<String, AttributeValue> attributes = Maps.newHashMap();
+    populateAttemptNumber(attributes);
+
+    span.addAnnotation("Attempt started", attributes);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void attemptSucceeded() {
+    Map<String, AttributeValue> attributes = baseAttemptAttributes();
+
+    span.addAnnotation("Attempt succeeded", attributes);
+  }
+
+  @Override
+  public void attemptCancelled() {
+    Map<String, AttributeValue> attributes = baseAttemptAttributes();
+
+    span.addAnnotation("Attempt cancelled", attributes);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void attemptFailed(Throwable error, Duration delay) {
+    Map<String, AttributeValue> attributes = baseAttemptAttributes();
+    attributes.put("delay ms", AttributeValue.longAttributeValue(delay.toMillis()));
+    populateError(attributes, error);
+
+    String msg = error != null ? "Attempt failed" : "Operation incomplete";
+    span.addAnnotation(msg + ", scheduling next attempt", attributes);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void attemptFailedRetriesExhausted(Throwable error) {
+    Map<String, AttributeValue> attributes = baseAttemptAttributes();
+    populateError(attributes, error);
+
+    span.addAnnotation("Attempts exhausted", attributes);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void attemptPermanentFailure(Throwable error) {
+    Map<String, AttributeValue> attributes = baseAttemptAttributes();
+    populateError(attributes, error);
+
+    span.addAnnotation("Attempt failed, error not retryable", attributes);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void responseReceived() {
+    attemptResponses++;
+    totalAttemptResponses++;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void requestSent() {
+    attemptRequests++;
```

Comment:
I added docs to describe it's thread safety. I think making it thread safe will add a lot of noise and affect performance. Also I don't think it's necessary because grpc is not really thread safe.